include/vcpkg/dependencies.h

```diff
@@ -40,11 +39,30 @@ namespace vcpkg
         EXCLUDED
     };
 
-    struct InstallPlanAction
+    struct BasicAction
     {
-        static bool compare_by_name(const InstallPlanAction* left, const InstallPlanAction* right);
+        static bool compare_by_name(const BasicAction* left, const BasicAction* right);
 
-        InstallPlanAction() noexcept;
+        PackageSpec spec;
+    };
+
+    struct PackageAction : BasicAction
+    {
+        PackageAction(PackageSpec spec, std::vector<PackageSpec> package_dependencies, InternalFeatureSet feature_list)
+            : BasicAction{std::move(spec)}
+            , package_dependencies(std::move(package_dependencies))
+            , feature_list(std::move(feature_list))
+        {
+        }
+
+        std::string displayname() const;
+
+        std::vector<PackageSpec> package_dependencies;
```

Comment:
Stupid question: Is there anything really all that bad that comes from giving all actions these members even if for Remove actions they are largely meaningless?

(I'm not saying I actually want that to be the case, I just think it's an interesting thought experiment)

Reply:
It either results in bugs or really messy code at all consumption sites:
1. Thinking that `std::vector<PackageSpec> package_dependencies;` is populated with the package's dependencies when it's really not
2. Having the member be `Optional<std::vector<PackageSpec>>` to clarify that it might not be filled, which then is dependent on the actual action type and requires all consumers to either `.value_or_exit(VCPKG_LINE_INFO)` (red flags in code review) or `*x.get()` (risking UB if the assumption changes).

The primary motivation for me behind splitting up the action types is to be able to get rid of the `Optional<>` wrappers inside `InstallPlanAction` (morally a variant).

Reply:
Sounds good!